infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/KubernetesInternalRuntime.java

```diff
@@ -203,6 +207,9 @@ protected void internalStart(Map<String, String> startOptions) throws Infrastruc
 
       startMachines();
 
+      previewUrlCommandProvisioner.provision(context.getEnvironment(), namespace);
```

Comment:
Provisioner is a suitable word since it provisions URL to commands but in this context, it may a bit confusing.
We used to use a provisioner for preparing the environment before the start, and in such a case phase where it should be run is different. Also, we have two provisioners for previewURL, it make me thinking about something like:
1. `PreviewUrlEndpointProvisioner`: provision service and ingress/route for preview URL. It does pretty the same as ServerExposer does but for endpoints. Maybe then rename it to `PreviewUrlExposer`?
2. `PreviewUrlCommandProvisioner`: it's similar to `ServerResolver` as well. If we take a look into `ServerExposer` `ServerResolver` interaction, we can see that `ServerExposer` provision needed Services,Ingresses along with some metadata about servers, and then `ServerResolver` grab this meta info and use, it allows to avoid double matching user-defined servers to services/ingress.
Maybe we could do something similar, `PreviewUrlEndpointProvisioner` provision annotation `org.eclipse.che.preview/run` - where run is a command name. Then PreviewUrlResolver once grabs all ingresses/routes and provisions preview into needed command. 

WDYT?

Reply:
1. done
2. I would need a bit more explanation. It should be annotation in what kind of object? What to store there? I guess we can't use same algorithm for name because there are 'no limitations' in command's name (length, special characters).

Reply:
I'm OK with leaving as is since the proposed way it not really trivial because of command name format.
And your code works fine!